externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala

```diff
@@ -114,6 +127,29 @@ case class SparkSQLEngine(spark: SparkSession) extends Serverable("SparkSQLEngin
     stopEngineExec.get.execute(stopTask)
   }
 
+  private[kyuubi] def startFastFailChecker(): Unit = {
+    val interval = conf.get(ENGINE_CHECK_INTERVAL)
+    val timeout = conf.get(ENGINE_SPARK_MAX_INITIAL_WAIT)
+    val failFastCheck: Runnable = () => {
+      if (!shutdown.get) {
+        while (backendService.sessionManager.getOpenSessionCount <= 0 &&
+          System.currentTimeMillis() - getStartTime < timeout) {
+          info(s"Waiting for the initial connection")
+          Thread.sleep(interval)
+        }
+        if (backendService.sessionManager.getOpenSessionCount <= 0) {
+          error(s"Spark engine has been terminated because no incoming connection" +
+            s" for more than $timeout ms, de-registering from engine discovery space.")
+          assert(currentEngine.isDefined)
+          currentEngine.get.stop()
+        }
+      }
+    }
+    failfastChecker =
+      Some(ThreadUtils.newDaemonFixedThreadPool(1, "spark-engine-failfast-checker"))
```

Comment:
We don't need a thread pool, let's port `ThreadUtils#runInNewThread` from Spark

Reply:
Got it, let me fix it.